orttraining/orttraining/training_ops/cuda/nn/dropout.h

```diff
@@ -21,5 +21,22 @@ class DropoutGrad final : public CudaKernel {
   const float default_ratio_;
 };
 
+template <typename T1, typename T2>
+class BiasDropout final : public CudaKernel {
+ public:
+  BiasDropout(const OpKernelInfo& info) : CudaKernel(info), default_ratio_(0.5) {
+    int64_t seed = 0;
+    if (info.GetAttr<int64_t>("seed", &seed).IsOK()) {
+      generator_ = onnxruntime::make_unique<PhiloxGenerator>(static_cast<uint64_t>(seed));
+    }
+  }
+
+  Status ComputeInternal(OpKernelContext* context) const override;
+
+ private:
+  mutable std::unique_ptr<PhiloxGenerator> generator_;
```

Comment:
>mutable std::unique_ptr<PhiloxGenerator> generator_ [](start = 2, length = 51)

ideally kernels should avoid mutable state, right? i wonder if we can put a random number generator into OpKernelContext

Reply:
Unfortunately, we have to make an exception for Dropout op... the random number generator is stateful... and each dropout need to have it's own generator to ensure reproducibility... 

So here's where we now stand

---
In reply to: [441077982](https://github.com/microsoft/onnxruntime/pull/4167#discussion_r441077982) [](ancestors = 441077982)